devito/mpi/routines.py

```diff
@@ -354,9 +354,9 @@ def _make_sendrecv(self, f, hse, key, **kwargs):
         count = reduce(mul, bufs.shape, 1)
         rrecv = MPIRequestObject(name='rrecv')
         rsend = MPIRequestObject(name='rsend')
-        recv = Call('MPI_Irecv', [bufs, count, Macro(dtype_to_mpitype(f.dtype)),
+        recv = GPUDirectMPICall('MPI_Irecv', [bufs, count, Macro(dtype_to_mpitype(f.dtype)),
```

Comment:
mpi.py should be free of "gpu" concepts.

You're on the right track, but I'd do it differently:

* Instead of `MPICall` and `GPUDirectMPICall`, I'd add `IrecvCall` and `IsendCall`, subclasses of `Call`, because after all that's all you need
* Avoid changing all mpi calls -- aside from `MPI_Irecv` and `MPI_Isend`, all can remain `Call`